src/expr/term_conversion_proof_generator.cpp

```diff
@@ -0,0 +1,246 @@
+                                         context::Context* c)
+    : d_proof(pnm, nullptr, c), d_rewriteMap(c ? c : &d_context)
+{
+}
+
+TConvProofGenerator::~TConvProofGenerator() {}
+
+void TConvProofGenerator::addRewriteStep(Node t, Node s, ProofGenerator* pg)
+{
+  // should not rewrite term more than once
+  Assert(!hasRewriteStep(t));
+  Node eq = t.eqNode(s);
+  d_proof.addLazyStep(eq, pg);
+  d_rewriteMap[t] = s;
+}
+
+void TConvProofGenerator::addRewriteStep(Node t, Node s, ProofStep ps)
+{
+  // should not rewrite term more than once
+  Assert(!hasRewriteStep(t));
+  Node eq = t.eqNode(s);
+  d_proof.addStep(eq, ps);
+  d_rewriteMap[t] = s;
+}
+
+void TConvProofGenerator::addRewriteStep(Node t,
+                                         Node s,
+                                         PfRule id,
+                                         const std::vector<Node>& children,
+                                         const std::vector<Node>& args)
+{
+  // should not rewrite term more than once
+  Assert(!hasRewriteStep(t));
+  Node eq = t.eqNode(s);
+  d_proof.addStep(eq, id, children, args);
+  d_rewriteMap[t] = s;
+}
+
+bool TConvProofGenerator::hasRewriteStep(Node t) const
+{
+  return !getRewriteStep(t).isNull();
+}
+
+std::shared_ptr<ProofNode> TConvProofGenerator::getProofFor(Node f)
+{
+  Trace("tconv-pf-gen") << "TConvProofGenerator::getProofFor: " << f
+                        << std::endl;
+  if (f.getKind() != EQUAL)
+  {
+    Trace("tconv-pf-gen") << "... fail, non-equality" << std::endl;
+    Assert(false);
+    return nullptr;
+  }
+  std::shared_ptr<ProofNode> pf = getProofForRewriting(f[0]);
+  if (pf == nullptr)
+  {
+    // failed to generate proof
+    Trace("tconv-pf-gen") << "...failed to get proof" << std::endl;
+    Assert(false);
+    return pf;
+  }
+  if (pf->getResult() != f)
+  {
+    Trace("tconv-pf-gen") << "...failed, mismatch: returned proof concludes "
+                          << pf->getResult() << std::endl;
+    Assert(false);
+    return nullptr;
+  }
+  Trace("tconv-pf-gen") << "... success" << std::endl;
+  return pf;
+}
+
+std::shared_ptr<ProofNode> TConvProofGenerator::getProofForRewriting(Node t)
+{
+  // we use the existing proofs
+  PRefProofGenerator prg(&d_proof);
+  LazyCDProof pf(d_proof.getManager(), &prg);
+  NodeManager* nm = NodeManager::currentNM();
+  // Invariant: if visited[t] = s or rewritten[t] = s and t,s are distinct,
+  // then pf is able to generate a proof of t=s.
+  // the final rewritten form of terms
+  std::unordered_map<TNode, Node, TNodeHashFunction> visited;
+  // the rewritten form of terms we have processed so far
+  std::unordered_map<TNode, Node, TNodeHashFunction> rewritten;
+  std::unordered_map<TNode, Node, TNodeHashFunction>::iterator it;
+  std::unordered_map<TNode, Node, TNodeHashFunction>::iterator itr;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(t);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    it = visited.find(cur);
+
```

Comment:
It'd be useful to say explicitly in the traversal code below the points that correspond to pre/post rewrites